spec/src/main/asciidoc/configprofile.asciidoc

```diff
@@ -0,0 +1,54 @@
+//
+// Copyright (c) 2020 Contributors to the Eclipse Foundation
+//
+// See the NOTICE file(s) distributed with this work for additional
+// information regarding copyright ownership.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// You may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// Contributors:
+// Emily Jiang
+
+
+[[configprofile]]
+== Config Profile
+
+Config Profile indicates the project phrase, such as dev, testing, live, etc. 
+=== Specify Config Profile
+
+The config profile can be specified via a system property `mp.config.profile`. It can be set when starting your microservice, e.g.
+[source, text]
+----
+java -jar myapp.jar -Dmp.config.profile=testing
+----
+
+=== How Config Profile works
+
+In supporting the Config Profile, the config property names will need to start with the naming convention of `%configProfile%`, (e.g. `%live%vechile.name`).
+
+For an instance, a config source can be specified as follows.
```

Comment:
Would it be possible to override profiles even if they are active?

Reply:
Do you mean during the running you could do `System.setProperty(mp.config.profile, "testing")`? I think this profile should only be looked up on microservice starting to avoid headaches.

Reply:
If you mean another config source with higher ordinal also contains the same config property `%live%vechile.name`, the value of `%live%vechile.name` will be taken from a configsource having a higher ordinal.

Reply:
In the case where you have a higher ordinal with just `vehicle.name`. Will the profile always take precedence no matter the ordinal?

Reply:
That's a trick question! :wink:

No existing implementation works that way.  Profile keys should only take precedence over configuration keys from the same or lesser configuration source otherwise users can have some truly surprising behavior.

Reply:
Yes, that is why I'm asking, so we can clarify it in the spec.

Reply:
> In the case where you have a higher ordinal with just vehicle.name. Will the profile always take precedence no matter the ordinal?
The profile should always take precedence no matter the ordinal because of the config name rules. If multiple config sources contain the same profile config (e.g. `%live%vehicle.name`), the config value from the highest ordinal wins.

Reply:
That's the opposite of how SmallRye and Quarkus work and it means that overriding config properties with system properties won't work if the original property is in a profile. Why would you do it this way? Do you know of some existing implementation that does things this way?

Reply:
Let's take an example of defining `%live%vehicle.name` in microprofile-config.properties and then in Kubernetes deployment file, Todd, a devops engineer, defines an environment variable '_live_vehicle_name` to override the value, I would expect the value from environment variable wins. By the way, as mentioned earlier, what we defined here should not be limited by the existing implementation.

Reply:
I agree with @Emily-Jiang that it's possible to overwrite properties in a profile with other properties in the same profile, no issue there. But the question is which of the approaches is better.

It's probably easier to maintain configuration if properties in a profile only override properties in the same config source. Otherwise it can be a huge mess to make sure that profiles are clearly defined.

On the other hand, there's a perfectly good usecase for overriding properties defined in all config sources. For example, if I have a custom config source that only defines properties in profiles. E.g., the default config is in a database and profile-specifc configs are in a file (or more files, e.g. one per profile).

I prefer to keep things simple and allow that profile properties override base properties only in the same config source. Or in all config sources with the same ordinal, although that might be awfully confusing.

Reply:
Overriding properties only in the same config source is also more efficient in the most cases. Here's some comparison of the number of requests to config sources for various cases (N is the number of config sources):

| Case        | across all config sources           | only in the same config source  |
| ------------- |:-------------:| -----:|
| config source with **highest** ordinal contains **profile** property      | 1 request | 1 request |
| source with **highest** ordinal contains **normal** property, only source with the **lowest** ordinal contains **profile** property       | 2xN      |   2 |
| source with the **highest** ordinal contains **normal** property, no source contains profile property | 2xN      |    2 |
| source with the **lowest** ordinal contains **profile** property, no other source contains profile or normal property | N      |    2xN |
| **no source** contains profile or normal property | 2xN      |    2xN | 
| only source with the **median** ordinal contains **profile** property, no other source contains profile or normal property  | N/2      |    N | 
| only source with the **median** ordinal contains **normal** property, no other source contains profile or normal property  | N + N/2      |    N | 

Mediam ordinal means that if you order sources from highest ordinal to lowest, the one in the middle has median ordinal.

Reply:
@OndroMih Thanks for the explaination and I understood what you meant. I think we are in agreement
profile properties override base properties only in the same config source. 
yields 
profile properties with a higher ordinal override base or other profile properties

Reply:
I'll update the spec wording to state "profile properties override base properties only in the same config source."